torch/lib/c10d/reducer.cpp

```diff
@@ -522,6 +594,17 @@ void Reducer::all_reduce_local_used_map() {
 }
 
 void Reducer::mark_variable_ready(VariableIndex index) {
+  // Rebuild bucket only if 1) it is the first time to rebuild bucket 2)
+  // find_unused_parameters_ is false, currently it does not support when
+  // there are unused parameters 3) this backward pass needs to run allreduce.
+  // Here, we just dump tensors and their parameter indices into
+  // rebuilt_params_ and rebuilt_param_indices_ based on gradient arriving
+  // order, and then at the end of finalize_backward(), buckets will be
+  // rebuilt based on rebuilt_params_ and rebuilt_param_indices_, and then
+  // will be broadcasted and initialized. Also we only need to dump tensors
+  // and parameter indices of one replica.
+  push_rebuilt_params(index);
```

Comment:
Wondering why we moved this from `autograd_hook` to mark_variable_ready?

Reply:
if grads are triggered more than once, this index will be put into rebuilt_params_ multiple times.  we only need to record this index once when this grad is really ready for communication.